// polar/Router.js
var Ctrl = require('./Ctrl').ctrl
var View = require('./View').View

var Router = {

  // Colection of app routes
  routes: [],

  // Adds route to memory. Expects routes to be names 'controller#action'
  add: function(route) {

    Router.routes.push(route)

  },

  // Checks if route is defined and passes control to controller.
  get: function(req, res) {

    Router.lookup(req, function(code, route) {

      if (code === 404) {

        res.writeHead(404, {'Content-Type': 'text/html'})
        View.render('public', '404', res)

      }

      if (code === 200) {

        Ctrl.find(route, req, res)

      }

    })
  },

  // Used in Router.get() to find route
  lookup: function(req, callback) {

    var reqUrl = req.url.split('/'),
        path = reqUrl[1],
        routeList = Router.routes

    for (var ii = 0; ii < routeList.length; ii++) {

      var route = Router._parseRoutes(routeList[ii]),
          matchedRoute;

      if (route.ctrlName === path) {

        matchedRoute = route

      }

    }

    if (matchedRoute) {

      return callback(200, matchedRoute)

    }

    return callback(404, matchedRoute)

  },


    // Parses routes
  _parseRoutes: function (route) {

    var ctrl        = route.split('#'),
        ctrlName    = ctrl[0],
        ctrlAction  = ctrl[1]

    return {
      ctrlName: ctrlName,
      ctrlAction: ctrlAction
    }
  }
}

exports.Router = Router
